### PR TITLE
Access module names

### PR DESCRIPTION
**Pull request by Brett208**
Finished a sprint towards incorporating #69.

Next step is probably adding similar functions to ini module loader.

I was pleasantly surprised that I could execute some limited unit tests against the ConsoleModuleLoader.

### CODE REVIEW

**Comment by Brett208** [comment]
This is based on PR #70. I haven't figured out how to not include the base branch's changes in the PR report.

File: srcStatic/ConsoleModuleLoader.h [hunk 1]
```diff
@@ -6,10 +6,14 @@
 class ConsoleModuleLoader {
 public:
 	ConsoleModuleLoader();
+	// Test Constructor
+	ConsoleModuleLoader(const std::string& testModuleDirectory);
```
Comment (DanRStevens) [inline srcStatic/ConsoleModuleLoader.h]: There's something about the comment that makes me want to remove it. I can understand why it's there though, and it probably should be there. Perhaps this relates to my earlier comments about dependency injection, and wondering if the constructors should be reworked such that a comment wouldn't be of value. Anyway, food for thought. No changes needed.

File: srcStatic/StringConversion.cpp [hunk 2]
```diff
@@ -47,3 +48,13 @@ std::size_t CopyStdStringIntoCharBuffer(const std::string& stringToCopy, char* b
 	// Unable to copy the whole string, so return needed buffer size
 	return stringToCopy.size() + 1;
 }
+
+void ToLowerInPlace(std::string& x) {
+	std::transform(x.begin(), x.end(), x.begin(), ::tolower);
+}
```
[the remark this answers is not in the document]
Comment (Brett208) [reply srcStatic/StringConversion.cpp]: Would this look like:

```C++
std::string& ToLowerInPlace(std::string& x) {
	std::transform(x.begin(), x.end(), x.begin(), ::tolower);

	return x;
}
```

File: srcStatic/ConsoleModuleLoader.cpp [hunk 3]
```diff
@@ -20,11 +20,29 @@ ConsoleModuleLoader::ConsoleModuleLoader()
 	moduleDirectory = FindModuleDirectory();
 }
 
+ConsoleModuleLoader::ConsoleModuleLoader(const std::string& testModuleDirectory)
+{
+	OutputDebugString("Console Module constructed in test mode.");
+
+	moduleDirectory = testModuleDirectory;
+}
```
[the remark this answers is not in the document]
Comment (Brett208) [reply srcStatic/ConsoleModuleLoader.cpp]: I am a fan of Dependency Injection. Although in the past I've taken it much too far, passing in 6-8 parameters into a constructor that didn't really contribute to testability.

As you mentioned, I think for the long term health of ConsoleModuleLoader, we should consider passing in the command line arguments and eliminate the empty constructor. There are some design changes to the overall structure of op2ext that may be required. I think maybe deferring this work to a more dedicated branch would be a better idea.

File: srcStatic/ConsoleModuleLoader.h (hunk 1 above)
Comment (Brett208) [reply srcStatic/ConsoleModuleLoader.h]: Once someone refactors ConsoleModuleLoader to only have 1 constructor, won't be an issue anymore. In its current state, should not be used outside test environment.

**Comment by Brett208** [comment]
Thanks for the quick review. Generally agree with all the suggestions. Need a bit more stick and rudder on how to implement ToLowerInPlace suggestion. I don't typically return references from functions.

File: srcStatic/StringConversion.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanRStevens) [reply srcStatic/StringConversion.cpp]: Yes, exactly.

**Comment by DanRStevens** [comment]
Your suggested implementation for returning a reference was exactly what was needed.

The one rule is to never return a reference to a local variable (or the address of a parameter, which is also part of the function's stack frame). In this case, the reference is passed in as a parameter value (not the address of the parameter). In particular, it was passed in by the caller, where you're returning it to. Naturally that address had better be valid for the caller, otherwise they shouldn't have been passing it to the function.

**Comment by Brett208** [comment]
Newlines at end of file fixed in commit to update ToLowerInPlace. Thanks for the help.

File: test/StringConversion.test.cpp [hunk 4]
```diff
@@ -11,11 +11,9 @@ TEST(StringConversion, ToLower)
 TEST(StringConversion, ToLowerInPlace)
 {
 	std::string str("TEST");
-	ToLowerInPlace(str);
-	EXPECT_EQ("test", str);
+	EXPECT_EQ("test", ToLowerInPlace(str));
```
Comment (DanRStevens) [inline test/StringConversion.test.cpp]: You may want to check both the return value and the original value is modified:
```cpp
EXPECT_EQ("test", ToLowerInPlace(str));
EXPECT_EQ("test", str);
```

Without both checks, the `ToLower` method could be used to satisfy the test, even though it offers different behavior from `ToLowerInPlace`.

File: test/StringConversion.test.cpp [hunk 5]
```diff
@@ -11,11 +11,9 @@ TEST(StringConversion, ToLower)
...
 
 	// Test Empty string
 	str = "";
-	ASSERT_NO_THROW(ToLowerInPlace(str));
-	EXPECT_EQ("", str);
+	EXPECT_EQ("", ToLowerInPlace(str));
```
Comment (DanRStevens) [inline test/StringConversion.test.cpp]: Ditto.

**Pull request merged**